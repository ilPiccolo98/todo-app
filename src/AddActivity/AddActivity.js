import { React, useState } from "react";
import { copyArray } from "../Utilities/Utilities";

const AddActivity = (props) => {
  const [name, setName] = useState("");
  const [description, setDescription] = useState("");
  const [status, setStatus] = useState(false);

  const handleChangeNameField = (e) => {
    e.preventDefault();
    setName(e.target.value);
  };

  const handleChangeDescriptionField = (e) => {
    e.preventDefault();
    setDescription(e.target.value);
  };

  const handleChangeStatusCheckbox = (e) => {
    setStatus(e.target.checked);
  };

  const generateNewActivityId = (activities) => {
    return activities[props.activities.length - 1].id + 1;
  };

  const addNewActivity = (activities, newActivity) => {
    activities.push(newActivity);
  };

  const createNewActivity = (id, name, description, status) => {
    return {
      id,
      name,
      description,
      status,
    };
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (name.length !== 0 && description.length !== 0) {
      const newId = generateNewActivityId(props.activities);
      const activitiesCopy = copyArray(props.activities);
      addNewActivity(
        activitiesCopy,
        createNewActivity(newId, name, description, status)
      );
      props.setActivities(activitiesCopy);
    }
  };
  return (
    <form onSubmit={handleSubmit}>
      <div>
        <label data-testid="name-label-add-activity" htmlFor="name">
          Name
        </label>
        <input
          data-testid="name-field-add-activity"
          id="name"
          name="name"
          placeholder="Name"
          onChange={handleChangeNameField}
          value={name}
        />
      </div>
      <div>
        <label
          data-testid="description-label-add-activity"
          htmlFor="description"
        >
          Description
        </label>
        <input
          data-testid="description-field-add-activity"
          id="description"
          name="description"
          placeholder="Description"
          onChange={handleChangeDescriptionField}
          value={description}
        />
      </div>
      <div>
        <label data-testid="status-label-add-activity" htmlFor="status">
          Status
        </label>
        <input
          data-testid="status-checkbox-add-activity"
          type="checkbox"
          id="status"
          name="status"
          onChange={handleChangeStatusCheckbox}
          checked={status}
        />
      </div>
      <div>
        <input
          data-testid="insert-button-add-activity"
          type="submit"
          id="submit"
          name="submit"
          value="Insert"
        />
      </div>
    </form>
  );
};

export default AddActivity;
