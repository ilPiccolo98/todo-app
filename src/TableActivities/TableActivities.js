import { React } from "react";
import Header from "./Header/Header";
import Body from "./Body/Body";

const TableActivities = (props) => {
  return (
    <div>
      <table>
        <tbody>
          <Header />
          <Body
            activities={props.activities}
            setActivities={props.setActivities}
          />
        </tbody>
      </table>
    </div>
  );
};

export default TableActivities;
