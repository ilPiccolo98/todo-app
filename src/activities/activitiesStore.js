import { configureStore } from "@reduxjs/toolkit";
import activitiesSlice from "./activitiesSlice";

const storeActivities = configureStore({
  reducer: {
    activities: activitiesSlice,
  },
});

export default storeActivities;
