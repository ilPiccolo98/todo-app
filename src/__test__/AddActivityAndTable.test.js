import React from "react";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { Provider } from "react-redux";
import AddActivity from "../AddActivity/AddActivity";
import TableActivities from "../TableActivities/TableActivities";
import activitiesStore from "../activities/activitiesStore";

const renderTableActivities = () => {
  return render(
    <Provider store={activitiesStore}>
      <TableActivities />
    </Provider>
  );
};

const renderAddActivity = () => {
  return render(
    <Provider store={activitiesStore}>
      <AddActivity />
    </Provider>
  );
};

const getAddTaskFormFields = (component) => {
  const nameFieldAddActivity = component.getByTestId("name-field-add-activity");
  const descriptionFieldAddActivity = component.getByTestId(
    "description-field-add-activity"
  );
  const statusCheckboxAddActivity = component.getByTestId(
    "status-checkbox-add-activity"
  );
  const insertButtonAddActivity = component.getByTestId(
    "insert-button-add-activity"
  );
  return {
    nameFieldAddActivity,
    descriptionFieldAddActivity,
    statusCheckboxAddActivity,
    insertButtonAddActivity,
  };
};

const getRowDataCells = (tableActivities, index) => {
  const table = tableActivities.getByTestId("table-body");
  const row = table.childNodes[index];
  const idCellData = row.childNodes[0];
  const nameCellData = row.childNodes[1];
  const descriptionCellData = row.childNodes[2];
  const statusCellData = row.childNodes[3];
  return {
    idCellData,
    nameCellData,
    descriptionCellData,
    statusCellData,
  };
};

describe("testing AddActivity and TableActivities together", () => {
  it("should add a the first activity into the table", () => {
    const componentAddActivity = renderAddActivity();
    const {
      nameFieldAddActivity,
      descriptionFieldAddActivity,
      statusCheckboxAddActivity,
      insertButtonAddActivity,
    } = getAddTaskFormFields(componentAddActivity);
    fireEvent.change(nameFieldAddActivity, {
      target: {
        value: "activity1",
      },
    });
    fireEvent.change(descriptionFieldAddActivity, {
      target: {
        value: "activity1",
      },
    });
    fireEvent.click(statusCheckboxAddActivity);
    const componentTable = renderTableActivities();
    fireEvent.click(insertButtonAddActivity);
    const { idCellData, nameCellData, descriptionCellData, statusCellData } =
      getRowDataCells(componentTable, 4);
    expect(idCellData.textContent).toBe("4");
    expect(nameCellData.textContent).toBe("activity1");
    expect(descriptionCellData.textContent).toBe("activity1");
    expect(statusCellData.textContent).toBe("true");
    expect(componentAddActivity).toMatchSnapshot();
    expect(componentTable).toMatchSnapshot();
  });
});
