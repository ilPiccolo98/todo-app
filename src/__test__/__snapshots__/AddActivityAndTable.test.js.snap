// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing AddActivity and TableActivities together should add a the first activity into the table 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div>
          <label
            data-testid="name-label-add-activity"
            for="name"
          >
            Name
          </label>
          <input
            data-testid="name-field-add-activity"
            id="name"
            name="name"
            placeholder="Name"
            value="activity1"
          />
        </div>
        <div>
          <label
            data-testid="description-label-add-activity"
            for="description"
          >
            Description
          </label>
          <input
            data-testid="description-field-add-activity"
            id="description"
            name="description"
            placeholder="Description"
            value="activity1"
          />
        </div>
        <div>
          <label
            data-testid="status-label-add-activity"
            for="status"
          >
            Status
          </label>
          <input
            data-testid="status-checkbox-add-activity"
            id="status"
            name="status"
            type="checkbox"
          />
        </div>
        <div>
          <input
            data-testid="insert-button-add-activity"
            id="submit"
            name="submit"
            type="submit"
            value="Insert"
          />
        </div>
      </form>
    </div>
    <div>
      <div>
        <table>
          <tbody
            data-testid="table-body"
          >
            <tr>
              <th>
                Id
              </th>
              <th>
                Name
              </th>
              <th>
                Description
              </th>
              <th>
                Status
              </th>
            </tr>
            <tr>
              <td
                data-testid="td-id-1"
              >
                1
              </td>
              <td
                data-testid="td-name-1"
              >
                shopping
              </td>
              <td
                data-testid="td-description-1"
              >
                buy some stuff
              </td>
              <td
                data-testid="td-status-1"
              >
                false
              </td>
            </tr>
            <tr>
              <td
                data-testid="td-id-2"
              >
                2
              </td>
              <td
                data-testid="td-name-2"
              >
                have fun with my dog
              </td>
              <td
                data-testid="td-description-2"
              >
                take my dog for a walk and have fun with it
              </td>
              <td
                data-testid="td-status-2"
              >
                true
              </td>
            </tr>
            <tr>
              <td
                data-testid="td-id-3"
              >
                3
              </td>
              <td
                data-testid="td-name-3"
              >
                work out
              </td>
              <td
                data-testid="td-description-3"
              >
                work out a little bit with weights and barbells
              </td>
              <td
                data-testid="td-status-3"
              >
                false
              </td>
            </tr>
            <tr>
              <td
                data-testid="td-id-4"
              >
                4
              </td>
              <td
                data-testid="td-name-4"
              >
                activity1
              </td>
              <td
                data-testid="td-description-4"
              >
                activity1
              </td>
              <td
                data-testid="td-status-4"
              >
                true
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <form>
      <div>
        <label
          data-testid="name-label-add-activity"
          for="name"
        >
          Name
        </label>
        <input
          data-testid="name-field-add-activity"
          id="name"
          name="name"
          placeholder="Name"
          value="activity1"
        />
      </div>
      <div>
        <label
          data-testid="description-label-add-activity"
          for="description"
        >
          Description
        </label>
        <input
          data-testid="description-field-add-activity"
          id="description"
          name="description"
          placeholder="Description"
          value="activity1"
        />
      </div>
      <div>
        <label
          data-testid="status-label-add-activity"
          for="status"
        >
          Status
        </label>
        <input
          data-testid="status-checkbox-add-activity"
          id="status"
          name="status"
          type="checkbox"
        />
      </div>
      <div>
        <input
          data-testid="insert-button-add-activity"
          id="submit"
          name="submit"
          type="submit"
          value="Insert"
        />
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`testing AddActivity and TableActivities together should add a the first activity into the table 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div>
          <label
            data-testid="name-label-add-activity"
            for="name"
          >
            Name
          </label>
          <input
            data-testid="name-field-add-activity"
            id="name"
            name="name"
            placeholder="Name"
            value="activity1"
          />
        </div>
        <div>
          <label
            data-testid="description-label-add-activity"
            for="description"
          >
            Description
          </label>
          <input
            data-testid="description-field-add-activity"
            id="description"
            name="description"
            placeholder="Description"
            value="activity1"
          />
        </div>
        <div>
          <label
            data-testid="status-label-add-activity"
            for="status"
          >
            Status
          </label>
          <input
            data-testid="status-checkbox-add-activity"
            id="status"
            name="status"
            type="checkbox"
          />
        </div>
        <div>
          <input
            data-testid="insert-button-add-activity"
            id="submit"
            name="submit"
            type="submit"
            value="Insert"
          />
        </div>
      </form>
    </div>
    <div>
      <div>
        <table>
          <tbody
            data-testid="table-body"
          >
            <tr>
              <th>
                Id
              </th>
              <th>
                Name
              </th>
              <th>
                Description
              </th>
              <th>
                Status
              </th>
            </tr>
            <tr>
              <td
                data-testid="td-id-1"
              >
                1
              </td>
              <td
                data-testid="td-name-1"
              >
                shopping
              </td>
              <td
                data-testid="td-description-1"
              >
                buy some stuff
              </td>
              <td
                data-testid="td-status-1"
              >
                false
              </td>
            </tr>
            <tr>
              <td
                data-testid="td-id-2"
              >
                2
              </td>
              <td
                data-testid="td-name-2"
              >
                have fun with my dog
              </td>
              <td
                data-testid="td-description-2"
              >
                take my dog for a walk and have fun with it
              </td>
              <td
                data-testid="td-status-2"
              >
                true
              </td>
            </tr>
            <tr>
              <td
                data-testid="td-id-3"
              >
                3
              </td>
              <td
                data-testid="td-name-3"
              >
                work out
              </td>
              <td
                data-testid="td-description-3"
              >
                work out a little bit with weights and barbells
              </td>
              <td
                data-testid="td-status-3"
              >
                false
              </td>
            </tr>
            <tr>
              <td
                data-testid="td-id-4"
              >
                4
              </td>
              <td
                data-testid="td-name-4"
              >
                activity1
              </td>
              <td
                data-testid="td-description-4"
              >
                activity1
              </td>
              <td
                data-testid="td-status-4"
              >
                true
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <table>
        <tbody
          data-testid="table-body"
        >
          <tr>
            <th>
              Id
            </th>
            <th>
              Name
            </th>
            <th>
              Description
            </th>
            <th>
              Status
            </th>
          </tr>
          <tr>
            <td
              data-testid="td-id-1"
            >
              1
            </td>
            <td
              data-testid="td-name-1"
            >
              shopping
            </td>
            <td
              data-testid="td-description-1"
            >
              buy some stuff
            </td>
            <td
              data-testid="td-status-1"
            >
              false
            </td>
          </tr>
          <tr>
            <td
              data-testid="td-id-2"
            >
              2
            </td>
            <td
              data-testid="td-name-2"
            >
              have fun with my dog
            </td>
            <td
              data-testid="td-description-2"
            >
              take my dog for a walk and have fun with it
            </td>
            <td
              data-testid="td-status-2"
            >
              true
            </td>
          </tr>
          <tr>
            <td
              data-testid="td-id-3"
            >
              3
            </td>
            <td
              data-testid="td-name-3"
            >
              work out
            </td>
            <td
              data-testid="td-description-3"
            >
              work out a little bit with weights and barbells
            </td>
            <td
              data-testid="td-status-3"
            >
              false
            </td>
          </tr>
          <tr>
            <td
              data-testid="td-id-4"
            >
              4
            </td>
            <td
              data-testid="td-name-4"
            >
              activity1
            </td>
            <td
              data-testid="td-description-4"
            >
              activity1
            </td>
            <td
              data-testid="td-status-4"
            >
              true
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
